### ISSUE TITLE
Failed connecting to the YubiKey when launching App -> PIV on Windows 10

### ISSUE THREAD

**Issue by nafetsreuab** [Yubico/yubikey-manager-qt#294]
YubiKey Manager (ykman) version: 1.2.3
How was it installed?: Via https://developers.yubico.com/yubikey-manager-qt/Releases/yubikey-manager-qt-latest-win64.exe
Operating system and version: Windows 10 PRO, 20H2
YubiKey model and version: Yubikey 5 NFC - Firmware 5.2.7

After launching YubiKey Manager (run as Administrator) and selecting Applications -> PIV i see:

**Failed connecting to the YubiKey. Make sure the application has the required permissions.**

I would expect to be able to manage Pin/Certs/... like in this screenshot:

https://www.yubico.com/wp-content/uploads/2021/01/YubiKeyManager_PIV_Screenshot.png

I tried several Keys (all 5 NFC). All show the same behavior.
I tried also other computers, a virtual machine and a laptop. No change. 
I tried the latest windows 10 21H2. No change.
The related windows services are all running:

SCardSvr
ScDeviceEnum

On windows console (started as administrator) the following is logged:

ykman.exe piv info
WARNING: PC/SC not available. Smart card protocols will not function.
Error: No eligiable connections are available ([<class 'yubikit.core.smartcard.SmartCardConnection'>]).

Linux has no issues with the stick and i can manage PIV related stuff.

Please clarify if this is a known issue or other software is requiried.

**Comment by fdennis** [comment]
Could you clarify if you have tried on multiple physical computers (not a VM)?

**Comment by fdennis** [comment]
Furthermore, could you go to "Device Manager" (on a Win computer where you have this issue) and check the contents of "Smart card readers" and "Smart card"? Do you have any items under there? If yes, could you share what items?

**Comment by bauerstefan** [comment]
> Could you clarify if you have tried on multiple physical computers (not a VM)?

Yes. Real (physical) Computers and virtual machines with same result.

**Comment by bauerstefan** [comment]
![grafik](https://user-images.githubusercontent.com/10959745/131385719-80e6d019-7919-4e5a-a6f2-d3b803a2a1b7.png)

![grafik](https://user-images.githubusercontent.com/10959745/131385793-9080b7e8-33ea-487e-8f34-2c7f0c2b1b23.png)

![grafik](https://user-images.githubusercontent.com/10959745/131385847-778f07f1-834e-4008-92ce-ba5febc4e246.png)

**Comment by bauerstefan** [comment]
![grafik](https://user-images.githubusercontent.com/10959745/131385992-d5944813-bc44-4210-94aa-1c849752d3d7.png)

![grafik](https://user-images.githubusercontent.com/10959745/131386028-71e39e92-4c87-4760-9f54-c9fc6463dd86.png)

**Comment by fdennis** [comment]
Does "WUDF" under "Smart card readers" disappear when you remove the YubiKey?
Generally, when inserting a YubiKey, you should see "WUDF" under "Smart card readers" but you should also have a "Smart cards" entry in Device Manager. Looks like you don't have this entry.

**Comment by uc-msft** [comment]
I have the same issue on Windows 11. I bought two keys - 5C NFC and 5 NFC. If I insert the 5 NFC then everything appears fine in the device manager (WUDF & the smart card shows up) and the key is detected in Yubikey Manager. Now, if I unplug the 5 NFC and insert the 5C NFC then it will appear briefly in the Yubikey Manager and after that the key disappears & the error message about permissions show up. Note that I am running the app as administrator.

**Comment by dainnilsson** [comment]
@uc-msft so in your case, is it always the 5C that is problematic? Is that one showing up in the Device Manager (both as a Smart Card Reader, and as a Smart Card)? Are you only having trouble with it when used directly after the 5 NFC, or always?

**Comment by uc-msft** [comment]
> @uc-msft so in your case, is it always the 5C that is problematic? Is that one showing up in the Device Manager (both as a Smart Card Reader, and as a Smart Card)? Are you only having trouble with it when used directly after the 5 NFC, or always?

Yes, it is the 5C that is problematic. 5C does show up briefly and the devices in Device Manager go to the disconnected state after few seconds. (If I use the 'show hidden devices' in Device Manager I can see the entries). I have trouble with the 5C always.  On another laptop with Windows 10, I can use both the keys just fine.

**Comment by dainnilsson** [comment]
@bauerstefan, @uc-msft, I'm not sure what is causing this issue. It doesn't seem like it is related to YubiKey Manager per se, as the problem appears already in the Device Manager. What is clear is that there's a problem with accessing the YubiKey over PC/SC (smart card). I would suggest you both open Support tickets here: https://support.yubico.com/hc/en-us/requests/new and reference this thread for details. Maybe they've seen this issue before and know what to do.

**Comment by Vink03** [comment]
Hi, 
I have the same problems as **@bauerstefan**. I installed YubiKey Minidriver. But when I insert YubiKey 5 NFC (Firmware: 5.4.3) in USB slot I can see only Smart card reader under Device Manager. There is no Smart Card device. I try this on few business computers which are joined to domain (win10 20H2). The same behavior on all of them.

On home computer (also win10 20H2) the same YubiKey works OK. I can see Smart card readers and also Smart Card device under Device Manager.

On all computers Smart Card and Smart Card Device Enumeration Service services are running, when I inset YubiKey in USB slot.

Because of that I am not able to enrolle certificate to YubiKey. I get an error, that Smart Card is read-only.

Any suggestions?

**Comment by nafetsreuab** [comment]
It seems that microsoft does not really like 3rd-party hardware anymore for its pro/enterprise-customers and is going to rule out all options to work with 3rd-party smartcards.

We skipped our deployment of yubikeys and used the virtual-smarcard-service with the TPM2.0 chip on mainboard.
This allows so simply upload and manages certificates in TPM. the virtual smartcard behaves almost like a real one (pin required..)

**Comment by Vink03** [comment]
Thank you for fast answer.
We just buy few YubiKeys for tests. If they are not compatible with microsoft we will have to find other solution also.

Is there anyone who have no problems with like this on domain computers?

**Comment by nafetsreuab** [comment]
On long term i guess microsoft will block all kind of 3rd party security/authentication devices in favor of MS authenticator app and windows helo. time to move to linux ;)

**Comment by Vink03** [comment]
So you think there is no solution for this problem with YubiKey right now?

**Comment by dainnilsson** [comment]
@Vink03 I would suggest that you contact Yubico Support (https://support.yubico.com/hc/en-us/requests/new) as well if you haven't done so already. I know that there are many people using YubiKeys on Windows for many different use cases, and I'd be surprised if that doesn't include domain computers, but it's not something I myself have any experience with or knowledge about.

**Comment by Vink03** [comment]
I already opened case on Yuico Support, but progress of case is very slow. So I asked also here if someone can help.

**Comment by Vink03** [comment]
Hi,
We prepared new PC (clean installation of Windows 10) and on this PC cert enrolment was successful.

**Comment by badiku** [comment]
no problem at win10, but same error on Windows Server 2022.
OTP and FIDO work, PIV not work.
Failed connecting to the YubiKey. Make sure the application has the required permissions.

there's a exclamation ! mark on Device Manager: Microsoft Usbccid Smart Cardreader UMDF2

Smart Card  in services.msc not started. it stop soon after I start it .

**Comment by yavko** [comment]
I have the same issue on Gnu/Linux, though Idk if I should be running with root permissions or not.

**Comment by asheroto** [comment]
Same issue here

**Comment by jdesai61** [comment]
I have a Windows 11 Pro machine - same issue - running SmartCard service does not fix it

**Comment by StefanD986** [comment]
Everyone who has the problem with exclamation mark and who has the device show up as `Microsoft Usbccid Smart Cardreader (UMDF2)` (the UMDF2 is important), this might solve the issue for you https://github.com/OpenSC/OpenSC/issues/2541#issuecomment-1285672570

**Comment by jdesai61** [comment]
My issue turned out to be that I was accessing the machine over Remote Desktop (because the machine is in the basement). Apparently, the SmartCard functionality works differently when machine is accessed using RDP.

**Comment by folfgirl** [comment]
I am having some similar problems when I put in my yubikey 5 NFC & try to go into PIV it stops me & says "Failed connecting to the YubiKey. Make sure the application has the required permissions." I am running arch Linux 6.0.7-arch1-1  yubikey manager Firmware: 5.4.3

**Comment by lilarcor** [comment]
same issue on mac client

**Comment by gitbock** [comment]
Had the same problem with the app. However, I was successful using the command line:
`PS C:\Program Files\Yubico\YubiKey Manager> .\ykman.exe piv certificates import 9a mycert.p12`

**Comment by gaffneyd4** [comment]
> I have the same issue on Gnu/Linux, though Idk if I should be running with root permissions or not.
> 
> EDIT: its a 5c nfc EDIT: I just needed to start a daemon, i fixed it

I believe the daemon referred to here is `pcscd`.

On Ubuntu 22.04.1 LTS (Jammy Jellyfish) (`cat /etc/os-release`) I experienced this error as well.

I fixed by doing:
```
sudo apt install pcscd
sudo systemctl enable pcscd
sudo systemctl start pcscd
```

Now I can access the `piv` application on the yubikey through yubikey-manager.

**Comment by andjo** [comment]
I had the same symptom with the GUI Manager (Applications -> PIV show connection failure, although already running as admin).

In my case it was caused by having loaded bad data in one of the slots using `yubico-piv-tool.exe`. 
I could identity the slot with `yubico-piv-tool.exe -a status` and resolve it with `yubico-piv-tool.exe -s 9c -a delete-certificate`.

After that the YubiKey Manager started to work as expected.

**Comment by mirao** [comment]
@gaffneyd4 I have the same issue in Ubuntu 22.10 with YubiKey 5 Nano. Your solution works, thanks.

**Comment by Tribux** [comment]
Same problem on Win10pro, nothing from above or https://github.com/OpenSC/OpenSC/issues/2541#issuecomment-1285672570 helped :/

**Comment by spiritualcrab** [comment]
> > I have the same issue on Gnu/Linux, though Idk if I should be running with root permissions or not.
> > EDIT: its a 5c nfc EDIT: I just needed to start a daemon, i fixed it
> 
> I believe the daemon referred to here is `pcscd`.
> 
> On Ubuntu 22.04.1 LTS (Jammy Jellyfish) (`cat /etc/os-release`) I experienced this error as well.
> 
> I fixed by doing:
> 
> ```
> sudo apt install pcscd
> sudo systemctl enable pcscd
> sudo systemctl start pcscd
> ```
> 
> Now I can access the `piv` application on the yubikey through yubikey-manager.

This solution worked for me in Ubuntu 22.04. Thank You.

**Comment by karamellpelle** [comment]
I found this thread after having problems with the `ykman` CLI tool. The systemctl enable/start solution also works on Arch Linux. Before enabling, `systemctl status pcscd.service` said 

```
     Loaded: loaded (/usr/lib/systemd/system/pcscd.service; indirect; present: disabled)
     Active: inactive (dead)
TriggeredBy: * pcscd.socket
```
(written not copied, the asterix should be a ring).

The function of _systemd_ `.socket`s is to listen for signals and load `.service`s[^1]. The status was "dead", so something had happened. `sudo journalctl -b -u pcscd.service` says: _Started/PC/SC Smart Card Daemon._, _pcscd.service: Referenced but unset environment variable evaluates to an empty string: PCSCD_ARGS_ . But empty PCSDC_ARGS is not the problem[^2]. The startup setting for `pcscd` in _/usr/lib/systemd/system/pcscd.service_ is 

```
ExecStart=/user/bin/pcscd --foreground --auto-exit $PCSCD_ARGS
ExecReload=/usr/bin/pcscd --hotplug
```
[The manual of _pcscd_](https://man.archlinux.org/man/pcscd.8) says for `--auto-exit`:
```
pcscd will quit after 60 seconds of inactivity after the release of the last PC/SC context. 
This is used when pcscd is started on demand by systemd.
```
Comparing the system log, the service is deactivated 64 seconds after start. 

I think `ExecReload` is triggered if you manually enable the service. [Arch Linux suggests service enable](https://wiki.archlinux.org/title/YubiKey). 

I think there is something wrong with the _pcscd_ socket for _systemd_: maybe it never trigger `ExecReload`. Manually enabling the service is probably not the "correct" solution since the socket should trigger the service.

[^1]: https://unix.stackexchange.com/questions/159462/what-is-systemds-target-service-and-socket
[^2]:  BTW, environment can be set in /etc/default/pcscd

**Comment by karamellpelle** [comment]
Live debugging above from me. I did `sudo systemctl enable pcscd.service`, which turned out to be wrong.  I've thought I had to explicitly specify `.service`, `.service` or `.target`.

`sudo systemctl enable pcscd` (without `.service`) enables `pcscd.socket`, not the service, which is the correct solution 😄
